crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -173,14 +177,34 @@ impl PgManager {
         .await
     }
 
-    async fn get_obj(
+    async fn get_obj(&self, address: Vec<u8>) -> Result<Option<StoredObject>, Error> {
+        self.run_query_async_with_cost(
+            move || Ok(QueryBuilder::get_obj(address.clone())),
+            |query| move |conn| query.get_result::<StoredObject>(conn).optional(),
+        )
+        .await
+    }
+
+    async fn get_obj_at_version(
         &self,
         address: Vec<u8>,
-        version: Option<i64>,
-    ) -> Result<Option<StoredObject>, Error> {
+        version: i64,
+    ) -> Result<Option<StoredHistoryObject>, Error> {
+        let (start, end) = self
+            .inner
+            .spawn_blocking(|this| this.get_consistent_read_range())
+            .await?;
+
         self.run_query_async_with_cost(
-            move || Ok(QueryBuilder::get_obj(address.clone(), version)),
-            |query| move |conn| query.get_result::<StoredObject>(conn).optional(),
+            move || {
+                Ok(QueryBuilder::get_obj_at_version(
+                    address.clone(),
+                    version,
+                    start,
+                    end,
+                ))
+            },
+            |query| move |conn| query.get_result::<StoredHistoryObject>(conn).optional(),
```

Comment:
This is not going to work, because it is not atomic -- the consistent range could have changed between the time that you issue the first query and the second query.  The queries need to be issued in one shot to the DB.

Reply:
Ah that's fair. 

There are two approaches here, one with subqueries, like
```
explain analyze SELECT *
FROM objects_history
WHERE checkpoint_sequence_number BETWEEN                                                                                           (SELECT MAX(checkpoint_sequence_number) FROM objects_snapshot)
    AND
    (SELECT MAX(sequence_number) FROM checkpoints)
AND object_id = '\x0000000000000000000000000000000000000000000000000000000000000006'
ORDER BY checkpoint_sequence_number, object_version DESC
LIMIT 1;
```

the cost and query planner result is a bit scary (it comes up with the brilliant idea to look through every single objects_history partition) 

Alternatively we could have a sql transaction that executes all three queries together (or two queries, I could union the query for the available range. Actually, this might be preferable because today, the query we make for the available range makes two separate transactions, one for the max checkpoint from snapshot for the left hand side, and one for the max checkpoint on checkpoints table for the right hand side.)

conn.build_transaction()
    .repeatable_read()
    .run(// queries)

I lean towards the latter because
1. It's hard to write subqueries in diesel
2. The query planning is suboptimal

We need to set the transaction to repeatable_read so that all reads within the transaction will see the same snapshot of the db. The default configured by diesel is read committed, where each statement within the transaction sees the database as it was at the start of that specific statement

Reply:
Building it as a transaction works for me.  I don't fully understand why the DB would treat it differently than if you used sub-queries, but hey if it works, it works!